### DISCUSSION TITLE
Accounting for movable stations in the import of .prices files

### DISCUSSION THREAD

**Discussion by eyeonus** [eyeonus/Trade-Dangerous#175] [General]
The solution I use in the listener is to list all movable stations as `MEGASHIP\{station name}` instead of '{system name}\{station name}` when creating the dict:
```
            # Mobile stations can move between systems. The mobile stations
            # have the following data in their entry in stations.jsonl:
            # "type_id":19,"type":"Megaship"
            # Except for that one Orbis station.
            # And now Fleet Carriers, they're type 24.
            if int(station['type_id']) == 19 or int(station['type_id']) == 24 or int(station['unq:station_id']) == 42041:
                full_name = "MEGASHIP"
            else:
                full_name = system_names[int(station['system_id@System.system_id'])]
            full_name += "/" + station['name'].upper()
            station_ids[full_name] = int(station['unq:station_id'])
```
That said, the listener also builds the dict at the start rather than as it encounters a 'new' station.

All fleet carriers have unique name in the form XXX-XXX. All rescue ships have "Rescue Ship" in their name. As far as I can tell, all the other megaships also have unique names as well.

**Comment by rmb4253** [comment]
Do rescue ships and megaships have markets? Presumably as they seem to appear in the prices file, yes. Could they be filtered out to avoid the problem?

**Comment by kfsone** [comment]
Aside, stupid optimization, given there are a lot of stations:

```
if int(station['type_id']) == 19 or int(station['type_id']) == 24 or int(station['unq:station_id']) == 42041:
```

`station['type_id']` requires the python virtual-machine to take the word 'type_id' and hash it, twice, then it has to look for the resulting entry in the station dict, each `station['type_id']` costing ~50ns. Add the int conversion, it comes out to 120-150ns for just those two matches.

Doing:
```
if int(station['type_id']) in (19, 24)
```
takes 70ns

better yet,
```
if station['type_id'] in ('19', '24')
```
takes 40ns

and if you could avoid unicode conversion and deal with the data in ascii/byte form:
```
if station['type_id'] in (b'19', b'24')
```

doing everything in byte form until we actually need to interact with text fields would probably be a huge perf win all over the place.

takes 35ns
